File: CSCI-1913/Labs/Lab2/altitools.py
'''
Authors: TODO: Zheng Robert Jia
         TODO: Daniyal Khan

'''
import math

def steepest_angle(lst):
    """accepts a list
    returns the steepest angle between two mountain ranges"""
    biggest_change = 0
    #Find the biggest change in height
    for i in range(1,len(lst)):
        print(biggest_change)
        dif = abs(lst[i]-lst[i-1])
        if dif > biggest_change:
            biggest_change = dif
    #Computer the angle in radians
    print(biggest_change)
    return math.degrees(math.atan(biggest_change))

def calc_distance(a,b):
    '''helper function
    accepts two altitude of horizontal distance 1
    return the distance between to points'''
    return math.sqrt((a-b)**2+1)

def total_distance(lst):
    '''accepts a list of the traveler's altitude
    returns the total distance the traveler needs to climb as a number'''
    # Check length
    if len(lst) <= 1:
        return 0
    distance = 0
    for i in range(len(lst)-1):
        distance += calc_distance(lst[i],lst[i+1])
    return distance

def longest_sequential_climb(lst):
    ''''accepts a list of traveler's altitude
    returns the longest sequential climb refer to the document for details'''
    # Make sure that the lst contains more than 1 element
    if len(lst) <= 1:
        return 0
    # store all the sequential climb in a list
    seq_climb = 0
    climb_rec = []
    # loop over the list
    # add up the numbers, and if lst[i]>lst[i-1] add the seq_climb and continue
    #                         else append the value to the sequential climb and make seq_climb 0
    for i in range(1,len(lst)):
        # determine if it is climb
        is_climb = (lst[i]-lst[i-1])>0
        climb = calc_distance(lst[i-1],lst[i])
        if is_climb:
            seq_climb += climb
        elif is_climb and i==len(lst)-1:
            seq_climb += climb
            climb_rec.append(seq_climb)
        else:
            climb_rec.append(seq_climb)
            seq_climb = 0
    return max(climb_rec)
    # return the maximum in the list

def num_of_peaks_and_valleys(lst):
    '''accepts a list
    returns a tuble (peak,valleys) representing
    the number of peaks and number of valleys'''
    # make sure the list has more than 2 elements
    if len(lst) <= 2:
        return (0, 0)
    valleys = 0
    peaks = 0
    # loop over the lists for peaks
    for i in range(1,len(lst)-1):
        if lst[i-1]<lst[i]>lst[i+1]:
            peaks += 1

    # loop over the lists for valleys
    for i in range(1, len(lst) - 1):
        if lst[i - 1] > lst[i] < lst[i + 1]:
            valleys += 1

    return peaks,valleys
def fill_valleys(inp,min_height):
    """accepts a list or tuple
    fills the valleys according to min_height
     return the new altitude"""
    # guarantee the input is changed into a list
    lst = list(inp)
    # enhanced for loop that changes every height below min_height into min_height
    for i in range(len(lst)):
        if lst[i] < min_height:
            lst[i] = min_height
    return lst

# print(math.degrees(math.atan(100)))

File: CSCI-1913/Labs/Lab6/src/Altitools.java
/*
 * CSCI 1913
 * Lab 06
 * Author: Zheng Robert Jia
 *          Daniyal Khan
 * */
// keep this line.
import java.lang.Math;
// this line will let you use function from the Java math package:
// details here: https://docs.oracle.com/en/java/javase/12/docs/api/java.base/java/lang/Math.html
// but a few functions you might need:
// Math.abs(number)
// Math.atan2(y, x)
// Math.pow(a,b) (which equals a^b)
// Math.toDegrees(radians)
// Math.sqrt(number)
// Use these roughly as written (so include the "Math." part)

public class Altitools {

    /**
     * Computes the steepest angle (Raising or falling) in a mountain range
     * @param mountainRange an array of doubles representing the heights of a mountain range. Will not be modified
     * @return the steepest upward/downward angle in the mountain range as a positive value measured in degrees.
     */
    public static double steepest_angle(double [] mountainRange) {
        double steepest_change = 0;
        for (int i = 1; i < mountainRange.length; i++) {
            double dif = Math.abs(mountainRange[i]-mountainRange[i-1]);
            if (dif > steepest_change){
                steepest_change = dif;
            }
        }
        return Math.toDegrees(Math.atan(steepest_change));
    }

    /**
     * Calculates the distance between the two mointain ranges 1 step interval apart
     * @param a first range
     * @param b second range
     * @return the distance between the two ranges
     */
    public static double calc_distances(double a, double b){
        return Math.sqrt(Math.pow((a-b),2)+1);
    }

    /**
     * Get the total "travel distance" over a mountain range
     * @param mountainRange an array of doubles representing the heights of a mountain range.  Will not be modified
     * @return the travel distance -- a non-negative double.
     */
    public static double total_distance(double [] mountainRange) {
        if (mountainRange.length <= 1){
            return 0;
        }
        double distance = 0;
        for (int i = 0; i < mountainRange.length-1; i++){
            distance += calc_distances(mountainRange[i],mountainRange[i+1]);
        }
        return distance;
    }

    /**
     * compute the longest (in terms of travel distance) seuqentially climbing part of a mountain range.
     * @param mountainRange an array of doubles representing the heights of a mountain range. Will not be modified
     * @return the latest travel distance of any sequentially climbing segment of the mountain range
     */
    public static double longest_sequential_climb(double [] mountainRange) {
        if (mountainRange.length <= 1){
            return 0;
        }
        double seq_climb = 0;
        boolean is_climb;
        double climb;
        double highest_climb = 0;
        for (int i = 1; i <mountainRange.length; i++){
            is_climb = (mountainRange[i] - mountainRange[i-1])>0;
            climb = calc_distances(mountainRange[i],mountainRange[i-1]);
            if (is_climb){
                seq_climb += climb;
                if (i==mountainRange.length-1 && seq_climb>highest_climb){
                    highest_climb = seq_climb;
                }
            }
            else if (seq_climb > highest_climb){
                highest_climb = seq_climb;
                seq_climb = 0;
            }
            else {
                seq_climb = 0;
            }
        }
        return highest_climb;
    }

    /**
     * compute the numebr of peaks (higher than it's neighbors) and valleys (lower than it's neighbors) in a mountain
     * @param mountainRange an array of doubles representing the heights of a mountain range. Will not be modified
     * @return a length 2 int array. position 1 is the number of peaks and position 2 the number of valleys.
     */
    public static int[] num_of_peaks_and_valleys(double [] mountainRange){
        if (mountainRange.length<2){
            return new int[]{0, 0};
        }
        int peaks = 0;
        int valleys = 0;

        // count peaks
        for (int i = 1; i<mountainRange.length-1; i++){
            if (mountainRange[i]>mountainRange[i-1] && mountainRange[i] > mountainRange[i+1]){
                peaks++;
            }
        }
        // count valleys
        for (int i = 1; i<mountainRange.length-1;i++){
            if (mountainRange[i]<mountainRange[i-1] && mountainRange[i]<mountainRange[i+1]){
                valleys++;
            }
        }
        return new int[]{peaks,valleys};
    }

    /**
     * Creates a new array to indicate what would happen if we were to hypothetically fill the lowest parts of a
     * mountain range up to a fixed height
     * @param mountainRange an array of doubles representing the heights of a mountain range. Will not be modified
     * @param target a double. the new array will have all positions below this filled up to this height
     * @return a new array showing the result of filling in the array
     */
    public static double[] fill_valleys(double [] mountainRange, double target){
        // copy the array first
        double[]copy_mountainRange = new double[mountainRange.length];
        for (int i = 0; i<mountainRange.length; i++){
            copy_mountainRange[i] = mountainRange[i];
        }
        // I could do enhanced for loop as well, but I don't know if I would lose points for that
        for (int i = 0; i<copy_mountainRange.length;i++){
            if(copy_mountainRange[i]<target){
                copy_mountainRange[i] = target;
            }
        }
        return copy_mountainRange;
    }

}

